[28기 김정현] [Refactor] 컴포넌트 분리 / 정적 데이터 / 네이밍 변경 등

KimJeongHyun opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 다양한 방식의 이미지 슬라이더 구현
- Intersection observer를 이용한 Infinite scroll 동작 구현
- 향후 BE 연결시 순조롭도록 Mockup data 설정 및 적절한 Map 함수의 사용
- scss 선택자에 익숙해지기
- Sticky 포지션 능숙하게 다루기

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 반응형 이미지 슬라이더 (MainFullSlider.js)
- 1626px 보다 작은 경우, transform:translateX(-slidePosition * 1447px)
- 1626px 보다 큰 경우, transform:translateX(-slidePosition * 1903px)
- 이미지 슬라이더 하단에는 현재 보고 있는 순서에 따라 dot width의 extend 또는 reduce 결정.
- setInterval 동작을 통해 자동 슬라이드 기능 적용됨.
- 테스트 방법 : Browser resizing 후 최대화 버튼 클릭하여 이미지가 크기에 맞게 떨어지는지 확인.

2. useScroll() 커스텀 훅 구현
- IntersectionObserver 사용
- fadeIn 구현되어 bestIcon, mdRecommend 영역에 적용됨.
- 현재 ref와 스크롤이 교차하는 경우에 대해서만 이벤트 동작 확인.
- 테스트 방법 : Scroll down시 bestIcon, mdRecommend 영역이 교차될 때 fadeIn 되는지 확인.

3. FlexibleSlider.js / Flex slider 영역 재사용 컴포넌트
- 2문단, 3문단에 적용됨.
- Ex) 6개의 슬라이드 데이터가 들어오는 경우 2 / 2 / 2개로 나뉘어 슬라이드됨.
- Ex) 3개의 슬라이드 데이터가 들어오는 경우 2 / 1 개로 나뉘어 슬라이드됨.
- 이미지 슬라이더 하단에는 현재 보고 있는 순서에 따라 dot width의 extend 또는 reduce 결정.
- 테스트 방법 : 버튼 클릭시 다음 순번의 이미지와 내용이 출력되고, 해당 순번의 dot이 길어지는지 확인.

4. MainStickyArea.js / Sticky 영역 재사용 컴포넌트
- 4,5,6 문단에 적용됨.
- 동일한 틀 안에서 전달받은 sectionArea에 따라 각각 다른 이미지와 내용을 렌더링함.
- 스크롤 위치에 따라 각 문단에 적용된 이벤트가 발생함.
- 현재 5문단 left, right 사용 중이나 transform으로 변경 예정.
- 테스트 방법 : 스크롤시 4, 6문단의 opacity와 scale이 변하는지 관측. 5문단의 신발 이미지의 위치가 변화하는지 관측.

5. Footer.js
- 반응형으로 작동함.
- min-width:1600px 일때 1140px
- min-width:1627px 일때 1627px
- 위 사항은 따로 screen mixin으로 구현할 예정이며, 프론트 팀원 논의가 필요한 사항이라 우선  static하게 해두었습니다.
- 테스트 방법 : 브라우저 크기 조절하여 width가 변경되는지 확인

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 이미지 슬라이더는 기능엔 크게 어려움이 없지만 정해진 규격이 없는 경우 반응형으로 img의 width와 transform을 건드려야 될 때 어떤 방법이 옳을까? 에 대한 고민이 있습니다. 과연 유저가 일부러 크기를 줄인 상태에서 이미지 슬라이드를 고려할까?를 생각했을 때 차라리 모바일/노트북/대형 환경으로 나눠 슬라이더를 보여주고 그것이 아니면 아예 사라지게 하는 게 맞지 않나? 란 생각을 하고 있습니다.
- className 네이밍시 직관적으로 하려고 하다보니 길이가 매우 길어지곤 합니다. 이 부분 신경쓰면서 Refactoring 진행할 생각입니다.
- 처음 커스텀 훅을 만들었고, state만의 변화로 스타일을 관리하는 것에 계속 익숙해지고 있는 듯 합니다.
- Backend에서 보내줄 데이터를 상상하며 데이터를 컨트롤할 수 있게끔 고민하면서 작성하고 있습니다.
- sticky position에 대한 이해가 부족해 꽤 헤맸습니다. 이 부분에 대한 블로그 포스팅 예정입니다.
- eventListener의 사용시 브라우저에 부담되지 않도록 만드는 방법에 대해 고민하며 작성하고 있습니다.
<br />

## :: 기타 질문 및 특이 사항
- 스크롤 업/다운을 통한 2개 이미지의 간격이 extend/reduce 되는 듯한 scroll event 구현 예정입니다. 
- Footer는 다소 중요성이 낮아 다음 주로 미룰 예정입니다.
- Backend data는 금주 금요일 또는 다음 주부터 붙여 테스트 가능할 것으로 생각됩니다.

KimJeongHyun committed:
[Add] Navbar 작성 완료

KimJeongHyun committed:
[Modify] Navigation component 위치 변경
 - Nav bar가 모든 page에 적용되는 바 router의 wrapper로 적용
 - 관련 태그 및 스타일 Router.js 및 index.scss로 이관

KimJeongHyun committed:
[Add] Main 1, 2문단 및 custom scroll hook 작성
 - 1문단 이미지 슬라이더 완료
 - 2문단 베스트 아이콘 레이아웃 및 슬라이더 완료
 - 커스텀 scroll fadein 활용을 통한 무한 슬라이더 구현
 - Main.js 컴포넌트 분리 진행
 - 전역 스크롤 이벤트 state 설정을 통해 과한 event 발생 방지(확인필요)

KimJeongHyun committed:
[Modify] Main 1문단 Image slider 반응형으로 변경
 - 1626px 우위 여부에 따라 img width 및 transform 로직 변경

KimJeongHyun committed:
[Modify] Main 1문단 image slider 반응형으로 변경 및 eslint 주의사항 수정
 - 1626px 우위 여부에 따라 image width, transform 변경 값 유동적으로
   변경

KimJeongHyun committed:
[Fix] main branch feature/Main 브랜치로 병합

KimJeongHyun committed:
[Modify] main branch 병합 후 변경 사항 정리
 - scss 변경에 따른 변수 값 변경 적용 후 테스트 완료

KimJeongHyun committed:
[Add] Main Slider 부분 수정
 - 공통 좌/우 버튼 컴포넌트화
 - 공통 Slider Dots Style mixin으로 통합
 - 1문단 Slider Image 부분 화면 꽉차도록 개편
 - 1문단 Slider Position에 따른 Slider Dot 변화 구현
 - 1문단 Slider Images Mockup data화 적용

KimJeongHyun committed:
[Refactor] scss 리팩토링
- 변수 이름 재정립
- nav 태그 정리
- scss 선택자 한 줄씩 띄우도록 정리
- strong 선택자 common으로 이관
- 기본 값 기재된 것들 삭제

KimJeongHyun committed:
[Fix] main branch 병합

KimJeongHyun committed:
[Modify] main merge 후 미반영 사항 정리
-index.scss에  headContainer style 반영

KimJeongHyun committed:
[Fix] 오 커밋된 페이지 롤백
- Main 부분 파일 삭제

KimJeongHyun committed:
[Refactor] Nav.scss 리팩토링 작업 및 Router.js 태그 수정
- Router.js 감싸는 태그  main에서 div로 변경
- Nav.js 불필요한 100% 값 일괄 삭제
- 공통 style mixin으로 통합
- 반응형으로 수정

KimJeongHyun committed:
[Modify] MainBestIcon 로직 변경 및 Router.js 소폭 수정
- Router.js main태그 div로 변경
- MainBestIcon.js 클릭 아이콘 mockdata 구조 설정
- MainBestIconContents.js 로직 오류 수정 (데이터가  3개 이하인 경우 오류 발생하여
  수정)
- MainBestIcon.js 클릭시 컨텐츠 영역 점멸하도록 트릭 추가

KimJeongHyun committed:
[Add] 3문단 MainStickyArea.js 작성 및 공통 버튼 컴포넌트화 작업 진행 등
- 3문단 scroll Event를 이용해 scale 및 opacity 조정 작업 진행. 코드 정리
  필요함.
- border button 컴포넌트화 진행. type을 입력하면 정해진 값에 의해 버튼이
  렌더링되도록 작성.
- Main.scss 속성 위치 변경 ( display > position > width.... else )
- border button 컴포넌트화에 따라 2문단 슬라이더 버튼 변경
- 1문단 슬라이더 dots 노란색으로 변경

KimJeongHyun committed:
[Refactor] MainStickyArea.js 코드 변경
- 스크롤 위치에 따른 분기 if 대신 boolean 판단으로 변경

KimJeongHyun committed:
[Modify] FlexSlider 컴포넌트화 작업 중

KimJeongHyun committed:
[Refactor] scroll throttle 적용, headContainer 삭제 등
- nav scroll에 throttle 적용.
- headContainer 삭제 (불필요한 것으로 판단함)
- Nav.scss 선택자 정리
- index.scss에서 headContainer 스타일 삭제
- common strong 선택자 정리

KimJeongHyun committed:
[Add] Flex Slider 재사용 컴포넌트화 및 side 작업 진행
- Main에 동일 구조 flex slider가 확인됨.
- 위의 사실에 따라 FlexibleSlider.js 재사용 컴포넌트 작성
- 공통된 틀 안에서 전달받은 sectionArea에 따라 다르게 detail 영역 구조를 변경함
- MainFlexSliderFetch는 전달받은 sectionArea와 selectedIcon(카테고리)에
  따라 slider에 표시될 이미지를 서버에 요청한다.
- MainFlexSlider는 전달받은 sectionArea에 맞는 레이아웃을 렌더링한다.
- MainSlider.js -> MainFullSlider로 네이밍 변경
- MD 추천 영역의 필요에 의해  BorderButton 마크업 및 작동 방식 변경
- MainBestIcon.js -> 재사용 컴포넌트화로 인해 삭제 처리
- Main.scss -> flexibleSlider mixin 추가 및 네이밍 변경에 따른 리팩토링
  진행

KimJeongHyun committed:
[Fix]재사용 btn에서 자세히 보기 btn인 경우 pointerEvent 활성화되도록 변경

KimJeongHyun committed:
[Modify] MainStickyArea 재사용 컴포넌트화 진행
- StickyArea는 sectionArea 값에 따라 다른 sticky 컨텐츠를 출력함
- 소개 -> 신발 소개 -> 비전 순서로 Main 재편성
- underSpan(밑줄 바) 및 buttonFill(채우기 버튼) mixin 변수화
- class black을 통해 위의 두 mixin 변수는 색상 전환이 가능함.

KimJeongHyun committed:
[Add] MainFullSlider setInterval에 따른 자동 슬라이드 구현

KimJeongHyun committed:
[Modify] MainStickyArea.js / 신발 부분 left 이동에서 transform으로 변경

KimJeongHyun committed:
[Add] FlexSlider 컨텐츠 점멸 기능 추가, BorderBtn fadeIn 기능 추가

KimJeongHyun committed:
[Refactor] import 위치 변경, 목데이터 분리,  3항 연산자 삭제 등
- import 위치 패키지 -> 컴포넌트  -> 목데이터 순서로 변경
- 임시 keyValue 데이터들 목데이터.js로 분기함
- && 연산자 활용을 통해 3항 연산자의 불필요한 return '' 제거

KimJeongHyun committed:
[Add] 하단 Join 부분 추가, Mixin 분리, topBtn 재사용 컴포넌트 생성 등
- 하단 Join 부분 별도 추가. Footer 요소가 아님.
- Mixin.scss 별도로 분기
- TopButton.js (클릭시 맨 위로 스크롤링) 추가
- className 리팩토링 ( && 에서 삼항 연산자로 변경 )

KimJeongHyun committed:
[Add] Footer.js 작업 완료
- Footer.js, Footer.scss 생성 및 구현
- Main에 Footer 추가 및 scss color 변수 변경

KimJeongHyun committed:
[Add] Footer.js 반응형으로 변경

KimJeongHyun committed:
[Add] Mockdata DB 대비 json으로 분기

KimJeongHyun committed:
[Refactor] 필요없는 import 삭제

KimJeongHyun committed:
[Add] Skeleton UI 적용 준비
- fetch 전, Skeleton UI에 해당되는 마크업이 렌더링됨.
- fetch 후, isLoad가 true가 되면 기존 UI가 렌더링됨.
- 현재는 점멸 수준이지만, DB 연결시 약간의 지연을 통해 효과가 있을
  것으로 예상됨.

saengmotmi commented:
@KimJeongHyun 여기도 충돌 해결해주시는데 이전 PR 머지하면 또 충돌 날 수도 있을거 같아서 순차적으로 해결해야 할거 같습니다
그리고 가급적 구현 화면은 PR에 포함시켜주시면 감사하겠습니다 🙇

saengmotmi commented on public/data/bestIcon1.json:
```diff
@@ -0,0 +1,9 @@
+    "id": 1,
+    "name": "고급진 신발",
+    "imgPath": "images/black_shoe.png",
+    "amount": "₩ 290,000",
```
number로 관리해주세요

KimJeongHyun replied:
Number로 받아온 뒤, 컴포넌트에서 원화표시 할 수 있도록 조정되었습니다!

saengmotmi commented on src/components/Buttons/Button.js:
```diff
@@ -0,0 +1,107 @@
+
+  const borderButtonScroll = useScrollFadeIn();
+
+  const borderBtnList = {
```
정적인 데이터라서 컴포넌트 바깥으로 분리하면 좋을거 같습니다

KimJeongHyun replied:
확인했습니다. 이 외에도 위와 같이 정적 데이터가 function에 숨어있는 경우를 찾아서 모두 컴포넌트 위쪽으로 올렸습니다.

saengmotmi commented on src/components/Buttons/Button.js:
```diff
@@ -0,0 +1,107 @@
+import { useScrollFadeIn } from '../../hooks/useScroll';
+
+export function MainSlideButton({ type, onClickMethod }) {
+  return (
+    <button name={type} onClick={onClickMethod}>
+      <span>{type.includes('Prev') ? `<` : `>`}</span>
+    </button>
+  );
+}
+
+export function MainBorderButton({
+  type,
```
지금은 버튼이 두개 뿐이지만 Buttons 폴더가 있는 만큼 각 컴포넌트 이름으로 별도 파일 분리하는게 좋겠습니다

KimJeongHyun replied:
타 페이지에서도 활용가능성이 있어보여 Main 접두어를 빼고, SliderButton.js와 BorderButton.js로 분리했습니다.

saengmotmi commented on src/components/Buttons/TopButton.js:
```diff
@@ -0,0 +1,36 @@
+    };
+  }, []);
+
+  const moveToTop = () => {
+    document.documentElement.scrollTop = 0;
+  };
```
scrollTo() 가 동작이라는 의미에 더 잘 맞는거 같습니다
https://developer.mozilla.org/ko/docs/Web/API/Window/scrollTo

KimJeongHyun replied:
확인했습니다! scrollTo(0,0)으로 동작 변경했습니다.

KimJeongHyun committed:
[Refactor] conflict 일괄 수정

saengmotmi commented on src/components/FlexibleSlider/FlexibleSlider.js:
```diff
@@ -0,0 +1,136 @@
+      <>
+        <img src={item.imgPath} alt="dummy" />
+        <img src={item.imgPath} alt="dummy" />
+        <div className="bestIconContentsEachDetail strong">
+          <div className="detailLeft">
+            <p>{item.name}</p>
+            <p>오리지널 | 블랙</p>
+            <span>
+              <FontAwesomeIcon icon={faHeart} />
+              {item.like}
+            </span>
+          </div>
+          <div className="detailRight">
+            <p>{item.amount}</p>
+            <MainBorderButton type="putCart" />
+          </div>
+        </div>
+      </>
+    );
+  };
+
+  const mdRecommendDetail = item => {
+    return (
+      <div className="mdRecommendDetail strong">
+        <p>{item.name}</p>
+        <span>
+          <FontAwesomeIcon icon={faHeart} />
+          {item.like}
+        </span>
+        <div className="flexSliderImages">
+          <img src={item.imgPath} alt="dummy" />
+          <img src={item.imgPath} alt="dummy" />
+        </div>
+        <div className="detailBox">
+          <p>{item.amount}</p>
+          <MainBorderButton type="putCart" />
+        </div>
+      </div>
+    );
+  };
```
컴포넌트는 가급적 별도 파일로 분리해주세요!

saengmotmi commented on src/pages/Main/MainFlexSlider.js:
```diff
@@ -0,0 +1,100 @@
+              </h3>
+            </div>
+          ))}
+        </div>
+      </section>
+    );
+  };
+
+  const mdRecommendBtns = () => {
+    return (
+      <section className="recommendBtnsWrapper">
+        <div className="recommendBtnsContainer" {...flexSliderMenuWrapper}>
+          <MainBorderButton
+            type="longBoots"
+            btnName="1"
+            selectedIcon={selectedIcon}
+            onClickMethod={stateHandler}
+          />
+          <MainBorderButton
+            type="collabo"
+            btnName="2"
+            selectedIcon={selectedIcon}
+            onClickMethod={stateHandler}
+          />
+          <MainBorderButton
+            type="furBoots"
+            btnName="3"
+            selectedIcon={selectedIcon}
+            onClickMethod={stateHandler}
+          />
+          <MainBorderButton
+            type="accGoods"
+            btnName="4"
+            selectedIcon={selectedIcon}
+            onClickMethod={stateHandler}
+          />
+        </div>
+      </section>
+    );
+  };
```
여기도 마찬가지입니다

KimJeongHyun replied:
컴포넌트 안에 컴포넌트를 또 넣어야되나.... 재사용 컴포넌트 한 개만으로 해결하고 싶었는데... 라는 고민이 좀 있었는데

종택님, 도현님 얘기듣고 미련없이 컴포넌트로 분리했습니다! 
지금 컴포넌트 / 하위 컴포넌트 / 하하위(?) 컴포넌트 이런 구조로 내려가는 모양새를 띄고 있는데
다음에는 물론 한 번 더 내려갈 순 있겠지만, 조금 더 깔끔한 구조로 만들어보겠습니다.

saengmotmi commented on src/pages/Main/MainFullSlider.js:
```diff
@@ -0,0 +1,102 @@
+  const [sliderPosition, setSliderPosition] = useState(1);
+  const [stopTran, setStopTran] = useState(false);
+  const [isWindowExtend, setIsWindowExtend] = useState(
+    window.innerWidth > 1626
```
의미가 있는 숫자일 경우 상수변수로 이름지어 분리해주세요

KimJeongHyun replied:
확인하여 컴포넌트 위 쪽으로 뺀 뒤 정상 작동 확인했습니다.
1626은 노트북 환경에서 적용되는 수치로 정해두었고, 프론트엔드분들 페이지가 어떻게 보이는지 확인하여 노트북 / 24인치 / 32인치 모니터 등 환경에 따라 다르게 가능하도록 config에서 꺼내온다던가.. 하는 방식으로 고려하고 있습니다.

saengmotmi commented on src/pages/Main/MainFullSlider.js:
```diff
@@ -0,0 +1,102 @@
+            <span
+              key={index}
+              className={index + 1 === sliderPosition ? 'active' : ''}
+              style={{ backgroundColor: 'yellow' }}
```
inline 스타일 사용하신 이유가 있을까요?

KimJeongHyun replied:
엑 FullSlider를 맨 처음에 했던 거라 그런가 있는지 까먹고 있었습니다 ㅠ.ㅠ className으로 제어하도록 변경했습니다.

KimJeongHyun commented:
@saengmotmi 
현재 메인 화면 동작은 주말까지 작업하여 백엔드 통신 제외 앞 단은 완료된 상태입니다! 
금일 작업 진행하면서 구현 화면 캡쳐하여 포함해두겠습니다. 확인 감사합니다.

saengmotmi reviewed (commented):
애니메이션이 있는 만큼 이 부분 역시 구현 화면이 PR에 추가되어 있으면 좋겠습니다

saengmotmi commented:
@KimJeongHyun 충돌 한번더 체크해주세요

KimJeongHyun committed:
[Refactor] 컴포넌트 분리 / 정적 데이터 / 네이밍 변경 등
- 정적 데이터 전역으로 처리
- FlexibleSlider 관련 컴포넌트들 일괄 분리 진행
- Button 컴포넌트 Border, Slider Button 으로 분리
- MockupData json 파일들 amount 값 Number로 변경
- MainFullSlider의 slideDot 색상 inline에서 class에서 제어하도록 변경
- Custom throttle function 작성 및 적용 테스트 완료
- TopButton 컴포넌트 scrollTop에서 scrollTo로 변

KimJeongHyun committed:
[Fix] Merge branch 'feature/Nav' into feature/Main

KimJeongHyun commented:
@saengmotmi 충돌은 해결했으나 common scss 관련하여, 스타일 변화가 의도대로 되지 않는 부분이 발생했습니다. 확인 후 comment 드리겠습니다.

KimJeongHyun committed:
[Fix] Conflict 수정 중 (1차) Common.scss 논의 중.

saengmotmi commented on src/components/Buttons/TopButton.js:
```diff
@@ -0,0 +1,36 @@
+import { useState, useEffect } from 'react';
+import { throttle } from 'lodash';
```
로대시 걷어내기로 했던거 같은데 이 부분 확인해주세요~

saengmotmi commented on src/styles/common.scss:
```diff
@@ -1,12 +1,23 @@
+  font-weight: 700;
+}
+
+* {
+  scroll-behavior: smooth;
+}
```
전체 선택자로 scroll-behavior 주는건 좀 어색해 보이는데 이거 확인 부탁드립니다

saengmotmi reviewed (commented):
고생하셨습니다~ 🔥
